File: xinceutong-web/api/_server/app/api/track.py
"""
api/track.py - 埋点事件接收

接收前端 /api/track 上报，存到 DB（可选），目前只打日志。
H5/小程序/App 都会调，schema 兼容。
"""
import json
from typing import Any, Dict

from fastapi import APIRouter, Request

from app.utils.logger import logger

router = APIRouter()


@router.post("")
async def track_event(request: Request) -> Dict[str, Any]:
    """埋点接收：开发期只 log，生产期可扩展写 DB / 发到分析平台"""
    try:
        body = await request.json()
    except Exception:
        body = {}

    event = body.get("event", "unknown")
    page = body.get("page", "-")
    platform = body.get("platform", "-")

    logger.info(f"[track] platform={platform} event={event} page={page} data={json.dumps(body.get('data') or {}, ensure_ascii=False)}")

    return {"ok": True}

File: xincetong-miniapp/src/utils/track.ts
/**
 * utils/track.ts - 埋点统计
 */
type TrackEvent =
  | 'page_view'
  | 'button_click'
  | 'assessment_start'
  | 'assessment_complete'
  | 'payment_success'
  | 'share_click'
  | 'grant_to_promoter'

interface TrackPayload {
  event: TrackEvent
  page?: string
  data?: Record<string, unknown>
}

/**
 * 平台检测（H5 build 时 __PLATFORM__ 不会替换成字面量，
 * 所以改用 uni.getSystemInfoSync() 运行时检测；H5 下 platform === 'web'，统一映射为 'h5'）
 */
function detectPlatform(): string {
  try {
    const sysInfo =
      typeof uni !== 'undefined' && uni.getSystemInfoSync
        ? uni.getSystemInfoSync()
        : null
    const p = sysInfo?.platform
    if (typeof p === 'string' && p.length > 0) {
      return p === 'web' ? 'h5' : p
    }
  } catch {
    /* ignore */
  }
  return 'h5'
}

export function track(payload: TrackPayload): void {
  // 开发期：仅打 log；生产期：调后端 /api/track 上报
  if (import.meta.env.DEV) {
    console.log('[track]', payload)
    return
  }
  // 静默失败，不打扰用户
  uni.request({
    url: ((import.meta.env.VITE_API_BASE_URL as string) || '') + '/api/track',
    method: 'POST',
    data: {
      ...payload,
      timestamp: Date.now(),
      platform: detectPlatform(),
    },
    fail: () => { /* 静默 */ },
  })
}

/** 页面 PV（onShow 调用） */
export function pageView(pageName: string, data?: Record<string, unknown>) {
  track({ event: 'page_view', page: pageName, data })
}
